test/controllers/users_controller_test.rb

```diff
@@ -272,7 +272,16 @@ class UsersControllerTest < ActionController::TestCase
     @request.session.clear
     @request.env['HTTP_REMOTE_USER'] = 'ares'
     get :extlogin
-    assert_redirected_to edit_user_path(User.unscoped.find_by_login('ares'))
+    assert_redirected_to ApplicationHelper.current_hosts_path
+  end
+
+  test "external user with email enabled but not set should be redirected to the edit page with warning" do
+    user = FactoryBot.create(:user, :mail_enabled => true, :auth_source_id => auth_sources(:one).id)
+
+    get :index, session: set_session_user.merge(:user => user.id)
+    assert_response :redirect
+    assert_redirected_to edit_user_path(user)
+    assert_equal "An email address is required, please update your account details", flash[:error]
```

Comment:
I agree that the test in its current form is essentially included in the other test. If we used extlogin in the test, we would simulate a first login of the external user, and their foreman account would be created at that moment. That means that the mail_enabled attribute could not exist prior to this moment and cannot be assigned to the account upon creation because the creation is automatic. At the same time, it is impossible to set the mail to enabled and at the same time leave email blank after the account creation, because this doesn't pass the validation. I realize I artificially create this scenario in the `#login shows a warning for any user model errors` test by stubbing the LDAP method return values, because admin can create an account with external authentication method, blank email, and mail enabled; but it is unclear to me whether an account created this way can actually be hooked up to an external login service later on. Stubbing the attributes here would essentially duplicate the test, as you previously pointed out. 
So, I propose to remove this test, since the scenario "Newly created external user with email enabled but not set" cannot occur.

Reply:
I'm still unclear on why we wouldn't include a similar test for external users, as we do for other user types.
If it's possible to create an external user with mail_enabled and an empty email address, then we should have a test case for that.

Reply:
Well, it is possible by either stubbing attributes as it is in the `#login shows a warning for any user model errors` test, or by using the Factory bot as it is now. However, I am not aware that it would be possible using `extlogin` as the test above uses. I should mention that how `extlogin` actually works goes beyond me. I don't see any user named 'ares' anywhere in the fixtures. Also, the second part of  `#login shows a warning for any user model errors`  already tests a scenario where user is external with mail enabled but not set.

Reply:
Deleted for now.